### PR TITLE
Add cancellation token parameter to async feature management interfaces.

### PR DESCRIPTION
**Pull request by jimmyca15**
This PR adds a cancellation token parameter to async feature management interfaces. Adding this parameter is a breaking change and requires a major version bump. The project versions have been updated to 3.0.0-preview.

### CODE REVIEW

File: src/Microsoft.FeatureManagement.AspNetCore/TagHelpers/FeatureTagHelper.cs [hunk 1]
```diff
@@ -55,8 +56,8 @@ public override async Task ProcessAsync(TagHelperContext context, TagHelperOutpu
                 IEnumerable<string> names = Name.Split(',').Select(n => n.Trim());
 
                 enabled = Requirement == RequirementType.All ?
-                    await names.All(async n => await _featureManager.IsEnabledAsync(n).ConfigureAwait(false)) :
-                    await names.Any(async n => await _featureManager.IsEnabledAsync(n).ConfigureAwait(false));
+                    await names.All(async n => await _featureManager.IsEnabledAsync(n, CancellationToken.None).ConfigureAwait(false)) :
+                    await names.Any(async n => await _featureManager.IsEnabledAsync(n, CancellationToken.None).ConfigureAwait(false));
             }
```
Comment (avanigupta) [inline src/Microsoft.FeatureManagement.AspNetCore/TagHelpers/FeatureTagHelper.cs:61]: Are we missing another `ConfigureAwait(false)` here?

File: build/NugetProperties.props [hunk 2]
```diff
@@ -6,10 +6,10 @@
 
   <!-- Official Version -->
   <PropertyGroup>
-    <MajorVersion>2</MajorVersion>
-    <MinorVersion>2</MinorVersion>
+    <MajorVersion>3</MajorVersion>
```
Comment (avanigupta) [inline build/NugetProperties.props:9]: I think you have some old code in this branch. The new versions now exist in .csproj files of both projects.

**Comment by avanigupta** [comment]
In the AspNetCore project, we have another public interface: `IDisabledFeaturesHandler`. Do we need to add CancellationToken for its `HandleDisabledFeatures` method?

**Comment by avanigupta** [comment]
>             TargetingContext targetingContext = await _contextAccessor.GetContextAsync().ConfigureAwait(false);

Does this `GetContextAsync` method need CancellationToken?

---
Refers to: src/Microsoft.FeatureManagement/Targeting/TargetingFilter.cs:52 in 8067cd6. [](commit_id = 8067cd623db037a37d545e7b434d2f7793a27594, deletion_comment = False)

File: src/Microsoft.FeatureManagement.AspNetCore/TagHelpers/FeatureTagHelper.cs (hunk 1 above)
Comment (jimmyca15) [reply src/Microsoft.FeatureManagement.AspNetCore/TagHelpers/FeatureTagHelper.cs:61]: Yes, thanks. I have added it.

File: build/NugetProperties.props (hunk 2 above)
Comment (jimmyca15) [reply build/NugetProperties.props]: You're right. I rebased and am now using the new versioning placement.

File: src/Microsoft.FeatureManagement.AspNetCore/Microsoft.FeatureManagement.AspNetCore.csproj [hunk 3]
```diff
@@ -3,9 +3,10 @@
 
   <!-- Official Version -->
   <PropertyGroup>
-    <MajorVersion>2</MajorVersion>
-    <MinorVersion>3</MinorVersion>
+    <MajorVersion>3</MajorVersion>
+    <MinorVersion>0</MinorVersion>
     <PatchVersion>0</PatchVersion>
+    <PreviewVersion>-preview</PreviewVersion>
```
Comment (avanigupta) [inline src/Microsoft.FeatureManagement.AspNetCore/Microsoft.FeatureManagement.AspNetCore.csproj:9]: We should only change the version of `Microsoft.FeatureManagement` package first. Once that has been released to NuGet, we can update the version in `Microsoft.FeatureManagement.AspNetCore`.

**Comment by jimmyca15** [comment]
> In the AspNetCore project, we have another public interface: `IDisabledFeaturesHandler`. Do we need to add CancellationToken for its `HandleDisabledFeatures` method?

It has ActionExecutingContext already which gives cancellation via HttpContext.RequestAborted. It is meant to be similar to [OnActionExecutionAsync ](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.filters.actionfilterattribute.onactionexecutionasync?view=aspnetcore-5.0)which behaves similarly.

File: src/Microsoft.FeatureManagement.AspNetCore/Microsoft.FeatureManagement.AspNetCore.csproj (hunk 3 above)
Comment (jimmyca15) [reply src/Microsoft.FeatureManagement.AspNetCore/Microsoft.FeatureManagement.AspNetCore.csproj]: Thanks, updated.

**Review by avanigupta — APPROVED** [review]

File: src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs [hunk 4]
```diff
@@ -40,7 +41,7 @@ public void Dispose()
             _changeSubscription = null;
         }
 
-        public Task<FeatureDefinition> GetFeatureDefinitionAsync(string featureName)
+        public Task<FeatureDefinition> GetFeatureDefinitionAsync(string featureName, CancellationToken _)
```
Comment (WeihanLi) [inline src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs:44]: Maybe better to have a proper parameter name like `cancellationToken`

File: src/Microsoft.FeatureManagement/IContextualFeatureFilter.cs [hunk 5]
```diff
@@ -18,7 +19,8 @@ public interface IContextualFeatureFilter<TContext> : IFeatureFilterMetadata
         /// </summary>
         /// <param name="featureFilterContext">A feature filter evaluation context that contains information that may be needed to evaluate the filter. This context includes configuration, if any, for this filter for the feature being evaluated.</param>
         /// <param name="appContext">A context defined by the application that is passed in to the feature management system to provide contextual information for evaluating a feature's state.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
         /// <returns>True if the filter's criteria has been met, false otherwise.</returns>
-        Task<bool> EvaluateAsync(FeatureFilterEvaluationContext featureFilterContext, TContext appContext);
+        Task<bool> EvaluateAsync(FeatureFilterEvaluationContext featureFilterContext, TContext appContext, CancellationToken cancellationToken);
```
Comment (WeihanLi) [inline src/Microsoft.FeatureManagement/IContextualFeatureFilter.cs:24]: Maybe keep a default value to simplify the operation when no need for cancelling?

File: src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs (hunk 4 above)
Comment (jimmyca15) [reply src/Microsoft.FeatureManagement/ConfigurationFeatureDefinitionProvider.cs:44]: I agree. I think the intent of using a discard here to indicate synchronous completion doesn't really fit will. I'll send a PR to update.

File: src/Microsoft.FeatureManagement/IContextualFeatureFilter.cs (hunk 5 above)
Comment (jimmyca15) [reply src/Microsoft.FeatureManagement/IContextualFeatureFilter.cs:24]: I think that's a good suggestion. Will add it to the PR to address the cancellation token parameter names.

**Pull request merged**